cryptography/src/bls12381/benches/dkg.rs

```diff
@@ -0,0 +1,142 @@
+        let dealers = Ordered::from_iter(private_keys.iter().map(|sk| sk.public_key()));
+
+        let (output, shares) = if previous {
+            let (o, s) = deal::<V, PublicKey>(&mut rng, dealers.iter().cloned()).unwrap();
+            (Some(o), Some(s))
+        } else {
+            (None, None)
+        };
+
+        let players = dealers.clone();
+        let info = Info::new(&[], 0, output, dealers, players).unwrap();
+
+        // Create player state for every participant
+        let mut player_states = private_keys
+            .iter()
+            .filter_map(|sk| {
+                let pk = sk.public_key();
+                if pk == me_pk {
+                    return None;
+                }
+                Some((
+                    pk,
+                    Player::<MinSig, PrivateKey>::new(info.clone(), sk.clone()).unwrap(),
+                ))
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        let mut logs = BTreeMap::new();
+        for sk in private_keys {
+            let pk = sk.public_key();
+            let (mut dealer, pub_msg, priv_msgs) = Dealer::start(
+                &mut rng,
+                info.clone(),
+                sk,
+                shares
+                    .as_ref()
+                    .and_then(|shares| shares.get_value(&pk).cloned()),
+            )
+            .unwrap();
+            for (target_pk, priv_msg) in priv_msgs {
+                // The only missing player should be ourselves.
+                if let Some(player) = player_states.get_mut(&target_pk) {
+                    if let Some(ack) = player.dealer_message(pk.clone(), pub_msg.clone(), priv_msg)
+                    {
+                        dealer.receive_player_ack(target_pk.clone(), ack).unwrap();
+                    }
+                }
+            }
+            logs.insert(pk, dealer.finalize().check(&info).unwrap().1);
+        }
+
+        Self { info, me, logs }
+    }
+
+    fn pre_finalize(
+        &self,
+    ) -> (
+        Player<V, PrivateKey>,
+        BTreeMap<PublicKey, DealerLog<V, PublicKey>>,
+    ) {
+        (
+            Player::<MinSig, PrivateKey>::new(self.info.clone(), self.me.clone()).unwrap(),
+            self.logs.clone(),
+        )
+    }
+}
+
+// Configure contributors based on context
+cfg_if::cfg_if! {
+    if #[cfg(full_bench)] {
+        const CONTRIBUTORS: &[u32] = &[5, 10, 20, 50, 100, 250, 500];
+        const CONCURRENCY: &[usize] = &[1, 4, 8];
+    } else {
+        const CONTRIBUTORS: &[u32] = &[5, 10, 20, 50];
+        const CONCURRENCY: &[usize] = &[1];
+    }
+}
+
+fn benchmark_dkg(c: &mut Criterion, previous: bool) {
+    let mut rng = StdRng::seed_from_u64(0);
+    for &n in CONTRIBUTORS {
+        let t = quorum(n);
+        let bench = Bench::new(&mut rng, previous, n);
+        for &concurrency in CONCURRENCY {
+            c.bench_function(
+                &format!("{}/n={} t={} conc={}", module_path!(), n, t, concurrency),
```

Comment:
If you take this route (makes sense), we need to add `previous` to this `format` line.